upup/pkg/fi/cloudup/urls.go

```diff
@@ -17,74 +17,162 @@ limitations under the License.
+		if err != nil {
+			return nil, nil, err
+		}
+
+		nodeUpHashURL, err := url.Parse(env + ".sha")
+		if err != nil {
+			return nil, nil, err
+		}
+		nodeUpLocation, nodeUpHash, err = assetsBuilder.RemapFileAndSHA(nodeUpLocation, nodeUpHashURL)
+		if err != nil {
+			return nil, nil, err
+		}
+		glog.Warningf("Using nodeup location from NODEUP_URL env var: %q", nodeUpLocation.String())
 	}
-	return nodeUpLocation
+
+	return nodeUpLocation, nodeUpHash, nil
 }
 
-// protokubeImageSource caches the ProtokubeImageSource value
-var protokubeImageSource string
+// protokubeLocation caches the protokubeLocation url
+var protokubeLocation *url.URL
+var protokubeHash *hashing.Hash
+
+// TODO FIXME make this a container when hosted assets
 
 // ProtokubeImageSource returns the source for the docker image for protokube.
 // Either a docker name (e.g. gcr.io/protokube:1.4), or a URL (https://...) in which case we download
 // the contents of the url and docker load it
-func ProtokubeImageSource() string {
-	if protokubeImageSource != "" {
+func ProtokubeImageSource(assetsBuilder *assets.AssetBuilder) (*url.URL, *hashing.Hash, error) {
+	if protokubeLocation != nil && protokubeHash != nil {
 		// Avoid repeated logging
-		return protokubeImageSource
+		glog.V(4).Infof("Using returning saved protokube location: %q", protokubeLocation)
+		return protokubeLocation, protokubeHash, nil
 	}
-	protokubeImageSource = os.Getenv("PROTOKUBE_IMAGE")
-	if protokubeImageSource == "" {
-		protokubeImageSource = BaseUrl() + "images/protokube.tar.gz"
-		glog.V(4).Infof("Using default protokube location: %q", protokubeImageSource)
+	env := os.Getenv("PROTOKUBE_IMAGE")
+	var err error
+	if env == "" {
+		protokubeLocation, protokubeHash, err = FileUrl("images/protokube.tar.gz", assetsBuilder)
+		if err != nil {
+			return nil, nil, err
+		}
+		glog.V(4).Infof("Using default protokube location: %q", protokubeLocation)
 	} else {
+		protokubeImageSource, err := url.Parse(env)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		protokubeHashURL, err := url.Parse(env + ".sha")
+		if err != nil {
+			return nil, nil, err
+		}
+		protokubeImageSource, protokubeHash, err = assetsBuilder.RemapFileAndSHA(protokubeImageSource, protokubeHashURL)
+		if err != nil {
+			return nil, nil, err
+		}
 		glog.Warningf("Using protokube location from PROTOKUBE_IMAGE env var: %q", protokubeImageSource)
 	}
-	return protokubeImageSource
+
+	return protokubeLocation, protokubeHash, nil
+}
+
+// FileUrl returns the base url for the distribution of kops - in particular for nodeup & docker images
+func FileUrl(file string, assetBuilder *assets.AssetBuilder) (*url.URL, *hashing.Hash, error) {
+	u, err := BaseUrl()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	u.Path = path.Join(u.Path, file)
+
+	sha, err := url.Parse(u.String())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sha.Path = path.Join(sha.Path, ".sha1")
+
+	fileUrl, hash, err := assetBuilder.RemapFileAndSHA(u, sha)
```

Comment:
`RemapFileAndSHA` looks like it is called the same way every time.  Maybe move the `.sha` concatenation & parsing into `RemapFileAndSHA` ?

Reply:
And the form in this function is more correct than the others, BTW

Reply:
I will refactor.

What do you mean form? More context please.